src/data/list/basic.lean

```diff
@@ -765,6 +772,14 @@ cons_head'_tail (head_mem_head' h)
 
 @[simp] theorem head'_map (f : α → β) (l) : head' (map f l) = (head' l).map f := by cases l; refl
 
+lemma tail_append_of_ne_nil (l l' : list α) (h : l ≠ []) :
+  (l ++ l').tail = l.tail ++ l' :=
+begin
+  cases l,
+  { simpa using h },
```

Comment:
Haven't tested this, but can't `contradiction` close a goal when there's a hypothesis `[] ≠ []`?
```suggestion
  { contradiction },
```

Reply:
I like the `simpa using h` syntax because it makes explicit which hypothesis is being used to prove the contradiction. This has two purposes -- `contradiction` has to check all available hypotheses, which could be slow; and it is more readable, especially in a large proof, which helps comprehension.

In this tiny proof, neither of those points are relevant. But in general, is there a style you prefer? I can, of course, switch to your suggestion. Just wanted to share why I phrased it this way in the first place.

Reply:
Well, conversely, `simpa` has to unify all `simp` lemmas involving `≠`, `=`, `≠` and `[]` to both the goal and `h` so without profiling I don't know which one will indeed turn out faster. I am pretty sure that `simpa` will produce a bigger proof term though.

My understandability argument in favour of using `contradiction` is that I read it as "this case never occurs, so focus on the remaining case(s)", while `simpa using h` reads more as "the conclusion follows from `h`", which could be used for a "relevant" alternative, so to say.

Reply:
Totally makes sense.